### PR TITLE
k8s/master Merger

### PR DESCRIPTION
**Pull request by stevenjcumming**
~**Note**: Merge https://github.com/department-of-veterans-affairs/vets-api/pull/16058 first~ merged

## Summary

Instead of updating the `k8s` branch and merging with `master`. This branch is updated with code from k8s and will be merged with master. 

- separate k8s build.yml workflows into multiple files (build.yml, code_checks.yml, deploy.yml)
- forward proxy cert moved from k8s to master
- remove datadog config for container agent
- move sidekiq_alive to initializer
- copy import-va-certs.sh from k8s
- minor updates on redis_store to match k8s

### Workflow Logic
1. Code Checks: trigger Code Checks on open PR and when pushed to master PRs
    - trigger Code Checks on open PR and when pushed to master PRs
2. Build:
    - do not build images and push to ECR on every open PR.
    - do build images after a PR is merged to master and merged code has completed Code Checks
3. Deploy:
    - do not deploy on every open PR.
    - merged PR should:
        - Pass Code Checks
        - Ensure the Build workflow has completed to ensure a docker image to push to the manifest repo

### Things not moved over from k8s branch
- .github/workflows/deploy-template.yml
- bin/deps
- bin/test
- bin/test-setup
- clamav_tmp/* (I believe these were committed accidentally)
- Documentation updates are in another PR that should be merged before this one
- **tmp/bundle_cache.tar.bz2** (I don't know what this is)

## Related issue(s)

- https://github.com/department-of-veterans-affairs/va.gov-team/issues/81621
- https://github.com/department-of-veterans-affairs/vets-api/pull/16058

## Testing done

- [ ] Testing master-based docker image on staging

## What areas of the site does it impact?
*don't need a `k8s` branch any more*

## Acceptance criteria

- [ ] Login on Review Instances
- [ ] Test `claim_letters` api endpoint on Review Instance
- [ ] Master is automatically deploy to non-prod environments
- [ ] Unit Tests, Linting, & Security Checks run on pushes
- [ ] Docker images are created/updated for PRs and master
- [ ] the `k8s` branch isn't needed for deployment

## Requested Feedback

*I'm not very familiar with the workflows so everything should be carefully reviewed*

### CODE REVIEW

File: lib/common/models/redis_store.rb [hunk 1]
```diff
@@ -39,7 +39,7 @@ def initialize(attributes = {}, persisted = false)
       begin
         super(attributes)
       rescue NoMethodError
-        Rails.logger.error('attributes failure: attributes')
+        Rails.logger.error("attributes failure: #{attributes}")
```
Comment (RachalCassity) [inline lib/common/models/redis_store.rb:42]: Good catch!

File: .github/workflows/deploy.yml [hunk 2]
```diff
@@ -0,0 +1,128 @@
...
+          ssm_parameter: /devops/github_actions_slack_bot_user_token
+          env_variable_name: SLACK_BOT_TOKEN
+      - name: slack notify
+        uses: ./.github/actions/vsp-github-actions/slack-socket
+        with:
+          slack_app_token: ${{ env.SLACK_APP_TOKEN }}
+          slack_bot_token: ${{ env.SLACK_BOT_TOKEN }}
+          message: "A packer build of ${{ matrix.build_location }} just failed."
+          blocks: "[{\"type\": \"divider\"}, {\"type\": \"section\", \"text\": { \"type\": \"mrkdwn\", \"text\": \":blob-disappointed: GitHub Action Runner Workflow ${{ matrix.build_location }} failed! :blob-disappointed:\n <https://github.com/${{ github.repository }}/actions/runs/${{ github.run_id }}|Workflow: ${{ github.workflow }} Run #${{ github.run_number }}>\"}}, {\"type\": \"divider\"}]"
+          channel_id: "C039HRTHXDH"
```
Comment (ericboehs) [inline .github/workflows/deploy.yml:128]: Missing new line.

File: .github/workflows/deploy.yml [hunk 3]
```diff
@@ -0,0 +1,128 @@
...
+          token: ${{ env.VA_VSP_BOT_GITHUB_TOKEN }}
+          fetch-depth: 1
+          path: vsp-infra-application-manifests
+          ref: refs/heads/main
+
+      - name: Update vets-api image and version name in Manifest repo for parent helm
+        env:
+          ECR_REGISTRY: ${{ steps.login-ecr.outputs.registry }}
+          ECR_REPOSITORY: dsva/vets-api
+        run: |
```
Comment (ericboehs) [inline .github/workflows/deploy.yml:68]: As soon as we merge this, deploys will start coming from master. I imagine we'll also want to simultaneously delete the workflows from the k8s branch.

File: import-va-certs.sh [hunk 4]
```diff
@@ -30,5 +33,5 @@ set -euo pipefail
 
     # Display VA Internal certificates that are now trusted
     awk -v cmd='openssl x509 -noout -subject' '/BEGIN/{close(cmd)};{print | cmd}' < /etc/ssl/certs/ca-certificates.crt \
-    | grep -i 'VA-Internal'
-)
+    | grep -iE '(VA-Internal|DigiCert)'
+)
```
Comment (ericboehs) [inline import-va-certs.sh:37]: Missing final new line.

**Review by ericboehs — CHANGES_REQUESTED** [review]
Can we use the deploy template like we're using in the k8s branch?

File: .github/workflows/deploy.yml [hunk 5]
```diff
@@ -0,0 +1,20 @@
+name: Deploy
```
Comment (RachalCassity) [inline .github/workflows/deploy.yml:1]: This should not be triggered until the Build workflow has successed.

File: .github/workflows/deploy.yml [hunk 6]
```diff
@@ -0,0 +1,20 @@
...
+on:
+  pull_request:
+    branches:
+      - master
+  push:
+    branches:
+      - master
+  workflow_call:
```
Comment (RachalCassity) [inline .github/workflows/deploy.yml:9]: workflow_call:
  workflow_dispatch:
  workflow_run:
    workflows: ["Build"]
    types: [completed]

File: .github/workflows/deploy.yml [hunk 7]
```diff
@@ -0,0 +1,18 @@
+name: Deploy
+on:
+  workflow_dispatch:
+  workflow_run:
+    workflows: ["Build"]
```
Comment (LindseySaari) [inline .github/workflows/deploy.yml:5]: I think the workflows: definition needs to change to the name of the build workflow, which I believe is ["Build & Push to ECR" here](https://github.com/department-of-veterans-affairs/vets-api/pull/16582/files#diff-5c3fa597431eda03ac3339ae6bf7f05e1a50d6fc7333679ec38e21b337cb6721R1)

File: config/ca-trust/fwdproxy.crt [hunk 8]
```diff
@@ -0,0 +1,24 @@
+-----BEGIN CERTIFICATE-----
```
Comment (LindseySaari) [inline config/ca-trust/fwdproxy.crt:1]: Is it ok to store this here or do we want it stored in param store>

File: import-va-certs.sh [hunk 9]
```diff
@@ -4,7 +4,10 @@ set -euo pipefail
 
 (
     cd /usr/local/share/ca-certificates/
-    
+
+    curl -LO https://cacerts.digicert.com/DigiCertTLSRSASHA2562020CA1-1.crt.pem
```
Comment (LindseySaari) [inline import-va-certs.sh:8]: Are these additional certs that we need on top of the ones we import from http://aia.pki.va.gov/PKI/AIA/VA/?

File: .github/workflows/code_checks.yml [hunk 10]
```diff
@@ -1,5 +1,11 @@
 name: Code Checks
-on: [push]
+on:
+  pull_request:
+    branches:
```
Comment (LindseySaari) [inline .github/workflows/code_checks.yml:4]: Do we want this to run on all branches, not just master? e.g. someone opens a PR against a branch that isn't master.

File: config/ca-trust/fwdproxy.crt (hunk 8 above)
Comment (stevenjcumming) [reply config/ca-trust/fwdproxy.crt:1]: @LindseySaari It was in the k8s branch so I moved it over. 

@ericboehs thoughts?

**Comment by stevenjcumming** [comment]
> @stevenjcumming I think we can move forward with the [workflow dispatch approach in the mockdata](https://github.com/department-of-veterans-affairs/vets-api-mockdata/blob/master/.github/workflows/workflow_dispatch.yml#L1-L4) repo (as detailed in the workflow comments). There may be some setup [involved here for the `vets-api-mockdata-update` event type](https://github.com/department-of-veterans-affairs/vets-api-mockdata/blob/master/.github/workflows/workflow_dispatch.yml#L32-L37)

@LindseySaari could this be a separate PR, I want to make sure I understand what you are saying before I change it.

File: .github/workflows/code_checks.yml [hunk 11]
```diff
@@ -1,5 +1,9 @@
 name: Code Checks
-on: [push]
```
Comment (RachalCassity) [inline .github/workflows/code_checks.yml:2]: `[](on: [push])` should stay

File: import-va-certs.sh (hunk 9 above)
Comment (RachalCassity) [reply import-va-certs.sh:8]: This is one of the certs from the SecretStore:` DigiCertTLSRSASHA2562020CA1.crt.pem` [here](https://github.com/department-of-veterans-affairs/vsp-infra-application-manifests/blob/main/apps/vets-api/prod/templates/secrets.yaml#L984)

Comment (RachalCassity) [reply import-va-certs.sh:8]: Is this file necessary?

Comment (RachalCassity) [reply import-va-certs.sh:8]: import-va-certs.sh matches the file in k8s.

File: config/initializers/datadog.rb [hunk 12]
```diff
@@ -6,10 +6,6 @@
 
 Datadog.configure do |c|
   if envs.include? Settings.vsp_environment
-    # Talk to DD agent in neighboring container
-    c.agent.host = 'datadog-agent'
-    c.agent.port = 8126
```
Comment (RachalCassity) [inline config/initializers/datadog.rb:11]: Datadog LGTM

File: import-va-certs.sh (hunk 9 above)
Comment (RachalCassity) [reply import-va-certs.sh:8]: @LindseySaari @stevenjcumming

File: .github/workflows/deploy-template.yml [hunk 13]
```diff
@@ -0,0 +1,99 @@
...
+
+      - name: Add and Commit file
+        uses: EndBug/add-and-commit@v9
+        with:
+          new_branch: main
+          add: "*"
+          cwd: vsp-infra-application-manifests/apps
+          author_name: va-vsp-bot
+          author_email: devops@va.gov
+          message: "Release ${{ needs.prepare-values.outputs.environments }} for ${{inputs.ecr_repository}}."
```
Comment (ericboehs) [inline .github/workflows/deploy-template.yml:99]: Missing new line at end of file 😄

File: import-va-certs.sh (hunk 9 above)
Comment (ericboehs) [reply import-va-certs.sh:8]: Kyle added these changes a while ago. It should be included in master (as this PR is doing).

File: config/ca-trust/fwdproxy.crt (hunk 8 above)
Comment (ericboehs) [reply config/ca-trust/fwdproxy.crt:1]: It [seems to be needed](https://github.com/department-of-veterans-affairs/va.gov-team/issues/41253) for the build process. 🤔

**Review by ericboehs — APPROVED** [review]

**Pull request merged**